car-bridge/src/streaming.rs

```diff
@@ -0,0 +1,222 @@
+            return Some(Action::Route(namespace, topic, source));
+        }
+
+        None
+    }
+}
+
+/// Tracks `EventProvider` instances by namespace.
+pub struct ProviderRegistry {
+    event_provider_by_namespace: HashMap<Namespace, Provider>,
+}
+
+impl ProviderRegistry {
+    pub fn new() -> Self {
+        Self { event_provider_by_namespace: HashMap::new() }
+    }
+
+    pub async fn register_event_provider(
+        &mut self,
+        chariott: &mut impl ChariottCommunication,
+        namespace: Namespace,
+    ) -> Result<(), Error> {
+        if let Entry::Vacant(e) = self.event_provider_by_namespace.entry(namespace.clone()) {
+            let event_provider = Provider::listen(chariott, namespace).await?;
+            e.insert(event_provider);
+        }
+
+        Ok(())
+    }
+
+    pub fn get_event_provider(&self, namespace: &Namespace) -> Option<&Provider> {
+        self.event_provider_by_namespace.get(namespace)
+    }
+
+    pub fn get_event_provider_mut(&mut self, namespace: &Namespace) -> Option<&mut Provider> {
+        self.event_provider_by_namespace.get_mut(namespace)
+    }
+}
+
+/// Represents events originating from a provider. This type allows distributing
+/// events coming from the provider to multiple consumers.
+pub struct Provider {
+    channel_id: String,
+    namespace: Namespace,
+    ess: Arc<EventSubSystem<Topic, Source, Event, Event>>,
+}
+
+impl Provider {
+    /// Listen to events originating from a provider.
+    pub async fn listen(
+        chariott: &mut impl ChariottCommunication,
+        namespace: Namespace,
+    ) -> Result<Self, Error> {
+        let (mut stream, channel_id) = chariott.open(namespace.clone()).await?;
+
+        let ess = Arc::new(EventSubSystem::new());
+
+        {
+            // Publish all subscribed values from a provider
+            // into its ESS.
+
+            let namespace = namespace.clone();
+            let ess = Arc::clone(&ess);
+
+            spawn(async move {
+                while let Some(event) = stream.next().await {
+                    let event = event.unwrap();
+                    ess.publish(event.source.clone().as_str(), event);
+                }
+
+                warn!("Stream for channel '{namespace}' broke.");
+            });
+        }
+
+        Ok(Self { ess, namespace, channel_id })
+    }
+
+    /// Links a topic to events coming from a provider. Prerequisite for routing
+    /// events.
+    pub fn link(&self, topic: Topic, response_sender: ResponseSender) {
+        let (_, mut topic_stream) = self.ess.read_events(topic.clone());
+
+        spawn(async move {
+            while let Some(e) = topic_stream.next().await {
+                let mut buffer = vec![];
+
+                if let Err(err) = e.encode(&mut buffer) {
+                    debug!("Failed to encode event: '{err:?}'.");
+                }
+
+                response_sender.send(Message::Event(buffer, topic.clone())).await;
```

Comment:
I don't think you want to send a potentially corrupted/partial message if encoding fails:

```suggestion
                let mut buffer = Vec::with_capacity(e.encoded_len());
                match e.encode(&mut buffer) {
                    Ok(_) => response_sender.send(Message::Event(buffer, topic.clone())).await,
                    Err(err) => debug!("Failed to encode event: '{err:?}'."),
                };
```